### ISSUE TITLE
AutoHeal recycles the process even if customer has configured LogEvent as an action (happens only when AutoAutoHeal is enabled)

### ISSUE THREAD

**Issue by puneetg1983** [ShekharGupta1988/AppLensV2#1226]
TLDR – The following statement inside “MONITORING_CONTEXT::DoWorkOnEndRequest”  in src4\hosting\processmonitoring\monitoringcontext.cpp  should be done under a lock because if multiple requests come, it is possible that one sets it to Recycle and the next request picks up the Recycle as UserDefinedAction before the first one ends up setting it back.
 
                // For Auto AutoHeal, always recycle.
m_Action = MONITORING_ACTION::Recycle;
 
Long Story 
I don’t know if I was lucky or unlucky but before making the new AutoHeal UI live, I thought let me test a few scenarios. So I created an AutoHeal rule like this, 
 
"autoHealRules": {
          "triggers": {
            "requests": {
              "count": 100,
              "timeInterval": "00:03:20"
            },
            "privateBytesInKB": 0,
            "statusCodes": [],
            "slowRequests": null
          },
          "actions": {
            "actionType": "LogEvent",
            "customAction": null,
            "minProcessExecutionTime": "00:00:00"
          }
        }
 
And I was expecting that on the 101th request, there should be an event in EventLog.xml but what I found that along with the event getting logged, the process was also getting recycled. I was running a stress test like this so I was definitely way more than 100 requests.
 
"c:\Program Files (x86)\IIS Resources\TinyGet\tinyget.exe" -srv:demowebcamp.azurewebsites.net -uri:/ -threads:50
"c:\Program Files (x86)\IIS Resources\TinyGet\tinyget.exe" -srv:demowebcamp.azurewebsites.net -uri:/ -threads:50
"c:\Program Files (x86)\IIS Resources\TinyGet\tinyget.exe" -srv:demowebcamp.azurewebsites.net -uri:/ -threads:50
"c:\Program Files (x86)\IIS Resources\TinyGet\tinyget.exe" -srv:demowebcamp.azurewebsites.net -uri:/ -threads:50
"c:\Program Files (x86)\IIS Resources\TinyGet\tinyget.exe" -srv:demowebcamp.azurewebsites.net -uri:/ -threads:50
 
First, I narrowed it down that it happens only if Proactive Auto-Healing is enabled. By setting WEBSITE_PROACTIVE_AUTOHEAL_ENABLED, I couldn’t repro the issue so made me think that it is due to Auto-AutoHealing only. I validated that in ApplicationHost.config file , the right value for monitoring was going
 
<monitoring>
        <triggers>
          <requests count="100" timeInterval="00:03:20" />
          <memory privateBytesInKB="0" />
        </triggers>
        <actions value="LogEvent" minProcessExecutionTime="00:00:00" />
      </monitoring>
 
So I debugged this and found that m_Action was getting set to Recycle which was causing the process to recycle. The only place where this happens is inside MONITORING_CONTEXT::DoWorkOnEndRequest so I created a breakpoint like this and ran the process from start
 
                bp ProcessMonitoring!MONITORING_CONTEXT::DoWorkOnEndRequest ".echo DoWorkOnEndRequst Hit;dt MONITORING_CONTEXT @ecx m_Action"
 
so we could see in the Debug log that around this time frame, the m_Action switched from Recycle to LogEvent
 
ProcessMonitoring!MONITORING_CONTEXT
   +0x168 m_Action : 1 ( LogEvent )
Debug session time: Wed Aug  1 12:15:05.993 2018 (UTC + 0:00)
System Uptime: 3 days 19:21:43.090
Process Uptime: 0 days 0:03:21.376
  Kernel time: 0 days 0:00:00.390
  User time: 0 days 0:00:00.875
DoWorkOnEndRequst Hit
ProcessMonitoring!MONITORING_CONTEXT
   +0x168 m_Action : 0 ( Recycle )
Debug session time: Wed Aug  1 12:15:06.008 2018 (UTC + 0:00)
System Uptime: 3 days 19:21:43.106
Process Uptime: 0 days 0:03:21.392
  Kernel time: 0 days 0:00:00.390
  User time: 0 days 0:00:00.875
DoWorkOnEndRequst Hit
 
And the only thing I could think this might be related to would be that we are setting it without protecting it. Please let me know if this makes sense and if my understanding is correct.
